docs: opensearch connector feature (#12998)
langflow-ai/langflow: docs/docs/Develop/knowledge.mdx

For more information on using knowledge bases in a flow, see the [**Knowledge Base** component](/knowledge-base) documentation.

### Configure vector database providers

> **DB Providers** are the vector store backend where knowledge base embeddings are stored and searched.

Reviewer: Optional. The plural UI label makes the sentence awkward.

```suggestion
**DB Providers** are the vector databases where your knowledge bases store and search embeddings.
```

> To configure DB Providers, go to **Settings → DB Providers**.

Reviewer: ```suggestion
To configure these providers, go to **Settings → DB Providers**.
```

The selected provider applies to all new knowledge bases you create.

> Existing knowledge bases keep their original backend.

Reviewer: Optional

```suggestion
Existing knowledge bases continue to use the provider that was active when they were created.
```

#### Chroma (default)

By default, knowledge bases use [ChromaDB](https://docs.trychroma.com/docs/overview/introduction) as a local vector store, with no additional setup required.
Knowledge bases are stored local to your Langflow instance.
The default storage location depends on your operating system and installation method:

- **Langflow Desktop**:
    - **macOS**: `/Users/<username>/.langflow/knowledge_bases`
    - **Windows**: `C:\Users\<name>\AppData\Roaming\com.LangflowDesktop\knowledge_bases`
- **Langflow OSS**:
    - **macOS/Windows/Linux/WSL with `uv pip install`**: `<path_to_venv>/lib/python3.12/site-packages/langflow/knowledge_bases` (Python version can vary. Knowledge bases aren't shared between virtual environments.)
    - **macOS/Windows/Linux/WSL with `git clone`**: `<path_to_clone>/src/backend/base/langflow/knowledge_bases`

If you set the `LANGFLOW_CONFIG_DIR` environment variable, the `knowledge_bases` subdirectory is created relative to that path.

To change the default `knowledge_bases` directory path, set the `LANGFLOW_KNOWLEDGE_BASES_DIR` environment variable:

```bash
export LANGFLOW_KNOWLEDGE_BASES_DIR="/path/to/parent/directory"
```

#### OpenSearch

> To use OpenSearch as a DB Provider, you need a running OpenSearch cluster accessible to your Langflow instance.

Reviewer: ```suggestion
To use OpenSearch as a database provider, you need a running OpenSearch cluster that is accessible to your Langflow instance.
```

> This example uses an OpenSearch container running locally, but you could also use a remote OpenSearch instance.

Reviewer: ```suggestion
This example uses an OpenSearch container running locally, but you can also use a remote OpenSearch instance.
```

> 1. Start a local OpenSearch container with security disabled. This allows you to connect without a username, password, or TLS.

Reviewer: ```suggestion
1. For this example, start a local OpenSearch container with security disabled. This allows you to connect without a username, password, or TLS. This configuration is for example purposes only; it _isn't_ recommended in production environments.
```

    ```bash
    podman run -d \
      --name opensearch \
      -p 9200:9200 \
      -p 9600:9600 \
      -e "discovery.type=single-node" \
      -e "plugins.security.disabled=true" \
      -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=YOUR_OPENSEARCH_PASSWORD" \
      opensearchproject/opensearch:latest
    ```

    :::note
    OpenSearch 3.x requires `OPENSEARCH_INITIAL_ADMIN_PASSWORD` to be set even when security is disabled.

    If the password fails validation, container startup exits immediately with `Password failed validation`.

>     Create a password that doesn't include `admin`, and contains uppercase and lowercase characters, a digit, and a special character.

Reviewer: In the OpenRAG docs, I just linked to the OpenSearch docs in case this criteria changes.

```suggestion
    The password must adhere to the https://docs.opensearch.org/latest/security/configuration/demo-configuration/#setting-up-a-custom-admin-password[OpenSearch password complexity requirements].
```

Although I don't see anything that says the password can't be `admin`. Maybe it's not documented.

    :::

2. Verify the cluster is reachable:

    ```bash
    curl -s http://localhost:9200
    ```

    A successful response indicates that the container has started and can receive requests:

    ```json
    {
      "name" : "your-node-name",
      "cluster_name" : "docker-cluster",
      "version" : {
        "distribution" : "opensearch",
        "number" : "3.6.0"
      },
      "tagline" : "The OpenSearch Project: https://opensearch.org/"
    }
    ```

>     If you get no response or a connection error, the container may still be starting. Wait a few seconds and try again.

Reviewer: ```suggestion
    If you get no response or a connection error, the container might still be starting. Wait a few seconds and try again.
```

    - **Username**: Leave blank if security is disabled. Otherwise, enter your basic auth username.
    - **Password**: Leave blank if security is disabled. Otherwise, enter your basic auth password.
    - **Default Index name**: Enter `langflow_knowledge`. The OpenSearch index to write and read from. This index is created in the later ingestion step, so it isn't immediately available.
    - **Vector field**: Enter `vector_field`. The document field for storing the embedding vector.
    - **Text field**: Enter `text`. The document field for storing the chunk text.
    - **Use TLS (HTTPS)**: Turn off. Enable if your cluster uses HTTPS.
    - **Verify TLS certificate**: Turn off. Enable if your cluster uses CA-signed certificates.

6. Click **Save and Use OpenSearch**.

    Optionally, click **Test Connection** to verify that Langflow can reach your OpenSearch cluster before saving.

    The OpenSearch database is now connected to Langflow as a knowledge base, so you can create a knowledge base that stores its embeddings in OpenSearch.

7. Click <Icon name="Library" aria-hidden="true"/> **Knowledge**, and then click <Icon name="Plus" aria-hidden="true"/> **Add Knowledge**.

> 8. Enter a name for this knowledge base. The name can be anything, and doesn't need to match the OpenSearch index name.

Reviewer: ```suggestion

8. Enter a name for this knowledge base. The name can be anything, and doesn't need to match the OpenSearch index name.
```

The name becomes the internal label used to scope searches to this knowledge base within the shared OpenSearch index.

> 9. Select an embedding model, and confirm **DB Provider** is set to **OpenSearch**.

Reviewer: ```suggestion

9. Select an embedding model, and confirm **DB Provider** is set to **OpenSearch**.
```

> 9. Select an embedding model, and confirm **DB Provider** is set to **OpenSearch**.

Reviewer: This step seems to contradict with the above: `The selected provider applies to all new knowledge bases you create.
Existing knowledge bases keep their original backend.`

If they can change the DB Provider at this step, then the previous lines can change to something like this maybe? `When you create a knowledge base in Langflow, you can choose one of your configured providers. Once you create a knowledge base, you cannot change its provider unless you recreate the knowledge base.`

>     Optionally, add **Custom Metadata Fields** to tag every chunk with additional context. For example, if you're ingesting files from multiple teams, add a field `team` with a value of `support`. When the **Knowledge Base** component searches, you can then filter results to only return chunks where `team` equals `support` to keep results scoped to the support team's content.

Reviewer: ```suggestion
10. Optional: Add **Custom Metadata Fields** to tag every chunk with additional context. For example, if you're ingesting files from multiple teams, add a field `team` with a value of `support`. When the **Knowledge Base** component searches, you can then filter results to only return chunks where `team` equals `support` to keep results scoped to the support team's content.

```

10. Click **Next Step**.
11. Add your source files and configure chunking settings, then click **Next Step**.

> 12. In the **Review & Build** pane, preview the first chunk of your data and confirm the chunk size is appropriate for your use case. A typical chunk size is 512–1000 characters. Smaller chunks give more granular retrieval but may lose context at boundaries. Click **Create**.

Reviewer: ```suggestion
11. Click **Next Step**.

12. Add your source files and configure chunking settings, then click **Next Step**.

13. In the **Review & Build** pane, preview the first chunk of your data and confirm the chunk size is appropriate for your use case. A typical chunk size is 512–1000 characters. Smaller chunks support more granular retrieval but they can lose context across chunks.

14. Click **Create**.
```
